dafni/main_dafni.py

```diff
@@ -0,0 +1,210 @@
+
+    parser.add_argument("--dag_path", required=True,
+                        help="Path to the input file containing a valid DAG (.dot). "
+                             "Note: this must be supplied if the --tests argument isn't provided.")
+
+    parser.add_argument('--output_path', required=False, help='Path to the output directory.')
+
+    parser.add_argument(
+        "-f",
+        default=False,
+        help="(Optional) Failure flag to step the framework from running if a test has failed.")
+
+    parser.add_argument(
+        "-w",
+        default=False,
+        help="(Optional) Specify to overwrite any existing output files. "
+             "This can lead to the loss of existing outputs if not "
+             "careful")
+
+    args = parser.parse_args(test_args)
+
+    # Convert these to Path objects for main()
+
+    args.variables_path = Path(args.variables_path)
+
+    args.tests_path = Path(args.tests_path)
+
+    if args.dag_path is not None:
+
+        args.dag_path = Path(args.dag_path)
+
+    if args.output_path is None:
+
+        args.output_path = "./outputs/results/"+"_".join([os.path.splitext(os.path.basename(x))[0]
+                                                          for x in args.data_path]) + "_results.json"
+
+        Path(args.output_path).parent.mkdir(exist_ok=True, parents=True)
+
+    else:
+
+        args.output_path = Path(args.output_path)
+
+        args.output_path.parent.mkdir(exist_ok=True, parents=True)
+
+    return args
+
+
+def read_variables(variables_path: Path) -> dict:
+    """
+    Function to read the variables.json file specified by the user
+    :param variables_path: A Path object of the user-specified file path
+    :returns:
+            - dict - A valid dictionary consisting of the causal tests
+    """
+    if not variables_path.exists() or variables_path.is_dir():
+
+        raise ValidationError(f"Cannot find a valid settings file at {variables_path.absolute()}.")
+
+    else:
+
+        with variables_path.open('r') as file:
+
+            inputs = json.load(file)
+
+        return inputs
+
+
+def validate_variables(data_dict: dict) -> tuple:
+    """
+    Function to validate the variables defined in the causal tests
+    :param data_dict: A dictionary consisting of the pre-defined variables for the causal tests
+    :returns:
+            - tuple - Tuple consisting of the inputs, outputs and constraints to pass into the modelling scenario
+    """
+    if data_dict["variables"]:
+
+        variables = data_dict["variables"]
+
+        inputs = [Input(variable["name"], eval(variable["datatype"])) for variable in variables if
+                  variable["typestring"] == "Input"]
+
+        outputs = [Output(variable["name"], eval(variable["datatype"])) for variable in variables if
+                   variable["typestring"] == "Output"]
+
+        constraints = set()
+
+        for variable, _inputs in zip(variables, inputs):
```

Comment:
Why is the element of `inputs` called `_inputs` rather than `input`? Typically the underscore convention is used for private variables

Reply:
@cwild-UoS ``input`` is already a pre-defined method in Python, so it's not a good idea to overwrite it! I don't think the variable name matters too much in this case, but I can change it if you think it's needed

Reply:
I think changing it from having an underscore would be good just to give off the right intent. Also isn't it already non pre-defined being `inputs` rather than `input`?